### DISCUSSION TITLE
How to find RabbitMQ version inside a mapped volume before RabbitMQ server starts?

### DISCUSSION THREAD

**Discussion by joe-agent** [docker-library/rabbitmq#727] [Q&A] [answered]
I have a custom script written in bash language. That script will run before `CMD ["rabbitmq-server"]` is called inside a custom RabbitMQ image container.

Where can I find the RabbitMQ version file within a RabbitMQ container?

**Comment by lukebakken** [comment, marked as answer]
Here's how I figured this out:

```
 docker run --name rabbitmq-727 --pull always  --rm -it -p 15672:15672 -p 5672:5672 rabbitmq:3-management
```

Then...

```
docker exec --interactive --tty rabbitmq-727 /bin/bash
```

Then, in the container, I ran `ps -ef`, and noted that RabbitMQ is installed to `/opt/rabbitmq`

```
root@5fc460a67742:/# ps -ef
UID          PID    PPID  C STIME TTY          TIME CMD
rabbitmq       1       0  0 15:50 pts/0    00:00:00 /bin/sh /opt/rabbitmq/sbin/rabbitmq-server
```

OK, fine, I know this container is running RabbitMQ 3.13.7, so maybe there is something with that version in the name:

```
root@cc2e562f3321:/opt/rabbitmq# find . -name '*3.13*'
...
...
./plugins/rabbit-3.13.7
...
...
```

OK, that means there will _always_ be a directory with the form `plugins/rabbit-VERSION`. You could use that.

Maybe there is a file with `3.13.7` in it. It's easy to check:

```
root@cc2e562f3321:/opt/rabbitmq# find . -type f -exec grep -F 3.13.7 {} +
```

There are quite a few files returned that have `3.13.7` in them, but this one is the most promising:

```
./plugins/rabbit-3.13.7/ebin/rabbit.app:        {vsn, "3.13.7"},
```

Aha, you can find the current RabbitMQ version by searching for `vsn` in the `/opt/rabbitmq/plugins/rabbit-*/ebin/rabbit.app` file. Done.

**Reply by michaelklishin** [reply]
Note that occasionally this mechanism won't be perfectly reliable https://github.com/rabbitmq/rabbitmq-server/issues/12339.

I'd use the package archive filename. Those by definition won't be incorrect or they will be retracted/never published.

**Comment by michaelklishin** [comment]
> Where can I find the RabbitMQ version file within a RabbitMQ container?

``` shell
rabbitmqctl version
```

will give you the CLI tools version but it is always the same as the RabbitMQ server version (unless you have multiple installations, that is).

**Reply by joe-agent** [reply]
Can `rabbitmqctl` command work before `CMD ["rabbitmq-server"]` runs?

**Reply by michaelklishin** [reply]
Yes. Some commands will require an explicit `--offline` but `version` is not one of them.

RabbitMQ must be installed before you run them but IAM sure that's obvious to you :)
